src/middleware/r2Middleware.ts

```diff
@@ -0,0 +1,138 @@
+import { CACHE_HEADERS } from '../constants/cache';
+import { Context } from '../context';
+import { R2Provider } from '../providers/r2Provider';
+import responses from '../responses';
+import { hasTrailingSlash, isDirectoryPath } from '../utils/path';
+import { Middleware } from './middleware';
+import { REDIRECT_MAP } from '../constants/r2Prefixes';
+import { Request } from '../routes/request';
+import { renderDirectoryListing } from '../utils/directoryListing';
+import { parseConditionalHeaders } from '../utils/request';
+
+// @ts-expect-error lazy loaded
+let provider: R2Provider = undefined;
+
+export class R2Middleware implements Middleware {
+  async handle(request: Request, ctx: Context): Promise<Response> {
+    if (provider === undefined) {
+      provider = new R2Provider({ ctx });
```

Comment:
Isn't the lifecycle of the execution of this file per request (serverless) or does it keep running over time? Cant we do a memoized version of this? i.e: a once() helper or something?

Reply:
By doing so you dont need to expose this global variable and access the provider variable in an unsafe manner on other functions.

So you set a:

`const getProvider = once((ctx) => new R2Provider({ ctx }));`

And then inside the functions simply call as `getProvider(ctx)`

Reply:
> Isn't the lifecycle of the execution of this file per request

If a request hits a colo that's already running an isolate with the worker it'll use that instead of spinning up another one iirc (within a 30s time constraint, cc @jasnell for fact checking).

Reply:
Right, we still can use the once approach imo